syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.registerallocationinference";
option java_outer_classname = "RegisterAllocationInferenceProto";
option objc_class_prefix = "REG";

package registerallocationinference;

// General Register Allocation Service 
service RegisterAllocationInference{

// RPC function to send Inference graph data and receive color data for codegen
//  rpc getColouring(Empty) returns (ColorData) {}
//rpc setGraph(GraphList) returns (Data) {}
  rpc getInfo(RegisterProfileList) returns (Data) {}
}

message GraphList {
  bytes payload=1;
}

message ColorData {
  bytes payload=1;
}

message RegisterProfileList {
  message RegisterProfile {
    uint32 regID=1;
    float spillWeight=2;
    repeated uint32 interferences=3;
    repeated uint32 useDistances=4;
    repeated float positionalSpillWeights=5;
  }
  repeated RegisterProfile regProf=1;
  bool result=2;
  bytes graph=3;
}

message Data {
  string message = 1;
  uint32 regidx = 2;
  int64 payload = 3;
  bytes colorData = 4;
}

message Empty {}
